### PR TITLE
Reworked rule generators

### PR DESCRIPTION
**Pull request by gkatsikas**
* Object-oriented design
* New OpenFlow rule generator element
* Bug fixes related to generated port numbers
* New example configurations

### CODE REVIEW

**Comment by tbarbette** [comment]
Nice! Those needed rework. I did a quick lookup and it's fine by me.

**Comment by gkatsikas** [comment]
I uncovered another bug apart. Once travis approves, I will merge

File: elements/ip/generateipfilterrules.cc [hunk 1]
```diff
@@ -154,30 +154,30 @@ GenerateIPFilter::configure(Vector<String> &conf, ErrorHandler *errh)
...
+GenerateIPFilterRules::simple_action(Packet *p)
 {
     // Create a flow signature for this packet
     IPFlowID flowid(p);
     IPFlow new_flow = IPFlow();
     new_flow.initialize(flowid & _mask);
 
     // Check if we already have such a flow
-    IPFlow *found = _map.find(flowid).get();
+    IPFlow *found = _map.find(flowid & _mask).get();
```
Comment (gkatsikas) [inline elements/ip/generateipfilterrules.cc:180]: Don'y you think that this was a bug?
Line 177 initializes a flow with a flow ID which derives from flowid & mask, but then line 180 searches for an existing flow with flowid. Apparently this is not the same flow

Comment (tbarbette) [reply elements/ip/generateipfilterrules.cc:180]: What is sure is that new_flow should be in the if branch. And then yes we need & _mask in both side. Probably flowid = flowid & _mask as we won't use it unmasked.

**Pull request merged**